fdbserver/DataDistribution.actor.cpp

```diff
@@ -973,51 +1174,215 @@ struct DDTeamCollection {
+			TraceEvent(SevWarn, "AddTeamWarning").detail("NotFoundMachineTeam", "OKIfTeamIsEmpty")
+				.detail("TeamInfo", teamInfo->getDesc());
+		}
+
+		bool healthy = true;
+		if ( teamInfo->serverIDs.size() != configuration.storageTeamSize ) { //A healthy team must have the desired number of servers. We may add an unhealthy team in initialization
+			healthy = false;
+		}
+
+		for( auto s = teamInfo->serverIDs.begin(); healthy && s != teamInfo->serverIDs.end(); ++s ) {
+			auto &status = server_status.get(*s);
+			if ( status.isFailed || status.isUndesired || status.isWrongConfiguration) {
+				healthy = false;
+			}
+		}
+		if ( healthy == false ) { //Print out the unhealthy team info.
+			int i = 0;
+			for ( auto s = teamInfo->serverIDs.begin(); s != teamInfo->serverIDs.end(); ++s ) {
+				auto &status = server_status.get(*s);
+				TraceEvent(SevInfo, "AddTeamAddUnhealthyTeam").detail("Primary", primary).detail("Index", i++)
+					.detail("ServerUID", s->first()).detail("IsFailed", status.isFailed)
+					.detail("IsUndesired", status.isUndesired).detail("IsWrongConfig", status.isWrongConfiguration);
+			}
+		}
+		teamInfo->machineTeam = machineTeamInfo;
 		teamInfo->tracker = teamTracker( this, teamInfo );
+		//assert( teamInfo->serverIDs.size() > 0 ); //team can be empty at DB initialization
 		if( teamInfo->servers.size() > configuration.storageTeamSize ) {
 			badTeams.push_back( teamInfo );
 		} else {
 			teams.push_back( teamInfo );
-			for (int i=0;i<newTeamServers.size();i++) {
+			for ( int i=0; i<newTeamServers.size(); i++ ) {
 				server_info[ newTeamServers[i]->id ]->teams.push_back( teamInfo );
 			}
 		}
 	}
 
-	template<class InputIt>
-	void addTeam( InputIt begin, InputIt end) {
-		vector< Reference<TCServerInfo> > newTeamServers;
-		for (auto i = begin; i != end; ++i) {
-			if (server_info.find(*i) != server_info.end()) {
-				newTeamServers.push_back(server_info[*i]);
-			}
+	void addTeam( const vector<Reference<TCServerInfo>>& newTeamServers, bool isInitialTeam ) {
+		//Reference<TCTeamInfo> teamInfo( new TCTeamInfo( newTeamServers ) );
+
+		vector<UID> servers;
+		for ( auto &server: newTeamServers ) {
+			servers.push_back(server->id);
 		}
+		addTeam(servers.begin(), servers.end(), isInitialTeam);
+	}
 
-		addTeam( newTeamServers );
+	template<class InputIt>
+	void addTeam( InputIt begin, InputIt end ) {
+		addTeam( begin, end, false);
 	}
 
 	void addTeam( std::set<UID> const& team ) {
-		addTeam(team.begin(), team.end());
+		addTeam(team.begin(), team.end(), false);
 	}
 
+	void addTeam( std::set<UID> const& team, bool isInitialTeam ) {
+		addTeam(team.begin(), team.end(), true);
+	}
+
+	/*
+	 * Add a machine team by using the machineIDs from begin to end
+	 */
+	Reference<TCMachineTeamInfo> addMachineTeam( vector<Standalone<StringRef>>::iterator begin, vector<Standalone<StringRef>>::iterator end) {
+		vector< Reference<TCMachineInfo> > machines;
+
+		for ( auto i = begin; i != end; ++i ) {
```

Comment:
I added another addMachineTeam() which takes a vector of TCMachineInfo reference to avoid the lookup when it's necessary.

In addTeam() where addMachineTeam() is called, we need to add machine team from machineIDs. We have to look up the TCMachineInfo reference anyway in that place. So I keep this addMachineTeam( vector<Standalone<StringRef>>::iterator begin, vector<Standalone<StringRef>>::iterator end), but change it to call the newly added addMachineTeam().